### DISCUSSION TITLE
Trouble accessing a binary in a crate

### DISCUSSION THREAD

**Discussion by ouillie** [bazelbuild/rules_rust#2786] [Q&A]
I need to use the `wit-bindgen` executable from [`wit-bindgen-cli`](https://docs.rs/crate/wit-bindgen-cli/latest) but I can't figure out why it won't work.

Here's an example in a single `MODULE.bazel` file:

```python
module(name = "foo")

bazel_dep(name = "rules_rust", version = "0.49.1")
crate = use_extension("@rules_rust//crate_universe:extension.bzl", "crate")

# For these 2 crates, I can get the bin
# at e.g. `@crates//:protoc-gen-prost__protoc-gen-prost`
crate.spec(package = "protoc-gen-prost", version = "0.3.1")
crate.annotation(crate = "protoc-gen-prost", gen_binaries = ["protoc-gen-prost"])
crate.spec(package = "protoc-gen-tonic", version = "0.4.0")
crate.annotation(crate = "protoc-gen-tonic", gen_binaries = ["protoc-gen-tonic"])

# For this one, I can't.
# Uncommenting the annotation causes an error.
crate.spec(package = "wit-bindgen-cli", version = "0.29.0")
#crate.annotation(crate = "wit-bindgen-cli", gen_binaries = ["wit-bindgen"])

crate.from_specs()
use_repo(crate, "crates")
```

Uncommenting the annotation for `wit-bindgen-cli` causes the error: `Error: Unused annotations were provided. Please remove them`. So, `rules_rust` seems to think `wit-bindgen-cli` [is not a dependency](https://github.com/bazelbuild/rules_rust/blob/f22926a2b4935ef4255ce1ecbce6d642862ac4d4/crate_universe/src/metadata/metadata_annotation.rs#L427) (verified with `bazel query`). Everything is fine for the first two. You can see all the generated targets with `bazel query '@crates//...' --output build`.

What is it about `wit-bindgen-cli` that makes it seemingly invisible to `rules_rust`?

**Comment by jgsogo** [comment]
I see the same error with `diesel_cli`:

```bzl
bazel_dep(name = "rules_rust", version = "0.57.1", dev_dependency = True)

crate = use_extension("@rules_rust//crate_universe:extension.bzl", "crate")

crate.spec(package = "diesel_cli", version = "2.2.7")
crate.annotation(crate = "diesel_cli", gen_binaries = ["diesel"])
crate.from_specs(name = "crates_diesel_cli")
use_repo(crate, "crates_diesel_cli")
```

Note.- My project also has a `crate.from_cargo` with its corresponding `Cargo.toml` file

**Comment by mikn** [comment]
I had similar issues with another binary called `tuftool`, but after I adding `artifact = "bin"` to the `.spec()` call these two to the `from_specs` call, it worked fine:
```
    generate_binaries = True,
    host_tools_repo = "rust_host_tools_nightly",
```

So for diesel_cli, it would be:
```python
crate.spec(artifact = "bin", package = "diesel_cli", version = "2.2.7")
crate.annotation(crate = "diesel_cli", gen_binaries = ["diesel"])
crate.from_specs(
    name = "crates_diesel_cli",
    generate_binaries = True,
    host_tools_repo = "rust_host_tools_nightly"
)
use_repo(crate, "crates_diesel_cli")
```

It is because they are binary crates rather than library crates. You can read more here: https://bazelbuild.github.io/rules_rust/crate_universe_bzlmod.html#binary-dependencies

**Comment by jgsogo** [comment]
Not working here 🤷  I also wrote in https://github.com/bazelbuild/rules_rust/issues/3110 as it's kind of related.

I get the following error:

```
bazel run @bindeps//:diesel_cli-diesel -- --help


error: failed to parse manifest at `/var/folders/jj/j4d2plrj28z03nwl751rr_2c0000gn/T/.tmpLNQfeQ/Cargo.toml`

Caused by:
  `artifact = …` requires `-Z bindeps` (diesel_cli)

Error: Failed to generate lockfile

Caused by:
    Failed to fetch crates for lockfile: exit status: 101
```

`bindeps` is unstable cargo-feature, but I don't know how to pass it to my build.

**Reply by mikn** [reply]
The key point is this line;
```
    host_tools_repo = "rust_host_tools_nightly"
```

Which makes `rules_rust` assemble the external repository using a nightly version of the toolchain, which has `bindeps` activated. (`bindeps` is an unstable feature that you cannot activate in the stable branch of the Rust toolchain yet)

I saw in your other comment that you register only a stable toolchain, either you don't register a toolchain (and `rules_rust` will pick the latest stable and nightly) or you need to register both a stable and a nightly toolchain (to specify a nightly toolchain also I think (!) you can add it to the `versions` list like this: `nightly-2025-02-17`. - you may also need to register a host toolchain if you register a regular one, but I am not enough of an expert to say for sure)

I opted for the first option in my testing.
